src/System.Private.CoreLib/shared/System/Text/AsciiUtility.cs

```diff
@@ -0,0 +1,1655 @@
+            }
+        }
+
+        /// <summary>
+        /// Given a DWORD which represents two packed chars in machine-endian order,
+        /// <see langword="true"/> iff the first char (in machine-endian order) is ASCII.
+        /// </summary>
+        /// <param name="value"></param>
+        /// <returns></returns>
+        private static bool FirstCharInUInt32IsAscii(uint value)
+        {
+            return (BitConverter.IsLittleEndian && (value & 0xFF80u) == 0)
+                || (!BitConverter.IsLittleEndian && (value & 0xFF800000u) == 0);
+        }
+
+        /// <summary>
+        /// Returns the index in <paramref name="buffer"/> where the first non-ASCII byte is found.
+        /// Returns <paramref name="bufferLength"/> if the buffer is empty or all-ASCII.
+        /// </summary>
+        /// <returns>An ASCII byte is defined as 0x00 - 0x7F, inclusive.</returns>
+        internal static unsafe nuint GetIndexOfFirstNonAsciiByte(ref byte buffer, nuint bufferLength)
+        {
+            // If SSE2 is supported, use those specific intrinsics instead of the generic vectorized
+            // code below. This has two benefits: (a) we can take advantage of specific instructions like
+            // pmovmskb which we know are optimized, and (b) we can avoid downclocking the processor while
+            // this method is running.
+
+            if (Sse2.IsSupported)
+            {
+                return GetIndexOfFirstNonAsciiByte_Sse2(ref buffer, bufferLength);
+            }
+
+            // Squirrel away the original buffer reference. This method works by determining the exact
+            // byte reference where non-ASCII data begins, so we need this base value to perform the
+            // final subtraction at the end of the method to get the index into the original buffer.
+
+            ref byte originalBuffer = ref buffer;
+
+            // Before we drain off byte-by-byte, try a generic vectorized loop.
+            // Only run the loop if we have at least two vectors we can pull out.
+            // Note use of SBYTE instead of BYTE below; we're using the two's-complement
+            // representation of negative integers to act as a surrogate for "is ASCII?".
+
+            if (Vector.IsHardwareAccelerated && bufferLength >= 2 * (uint)Vector<sbyte>.Count)
+            {
+                uint SizeOfVectorInBytes = (uint)Unsafe.SizeOf<Vector<sbyte>>(); // JIT will make this a const
+
+                if (Vector.GreaterThanOrEqualAll(Unsafe.ReadUnaligned<Vector<sbyte>>(ref buffer), Vector<sbyte>.Zero))
+                {
+                    // The first several elements of the input buffer were ASCII.
+                    // Pin and perform aligned reads until we find non-ASCII data or we hit the end of the buffer.
+                    // It's possible we'll end up reading the first part of the buffer twice; this is ok.
+
+                    buffer = ref Unsafe.AddByteOffset(ref buffer, SizeOfVectorInBytes);
```

Comment:
Are we gaining anything by writing half of this using the tricky `byref` pointer arithmetic?

I think all this should be using pinned pointers. It won't be any slower.

Reply:
I'll run the numbers on this in an hour or so.